runtime/nvqir/stim/StimCircuitSimulator.cpp

```diff
@@ -0,0 +1,192 @@
+ *                                                                             *
+ * This source code and the accompanying materials are made available under    *
+ * the terms of the Apache License 2.0 which accompanies this distribution.    *
+ ******************************************************************************/
+
+#include "nvqir/CircuitSimulator.h"
+#include "nvqir/Gates.h"
+#include "stim.h"
+
+#include <bit>
+#include <iostream>
+#include <set>
+#include <span>
+
+using namespace cudaq;
+
+namespace nvqir {
+
+/// @brief The StimCircuitSimulator implements the CircuitSimulator
+/// base class to provide a simulator delegating to the Stim library from
+/// https://github.com/quantumlib/Stim.
+class StimCircuitSimulator : public nvqir::CircuitSimulatorBase<double> {
+protected:
+  stim::Circuit stimCircuit;
+  std::mt19937_64 randomEngine;
+
+  /// @brief Grow the state vector by one qubit.
+  void addQubitToState() override { addQubitsToState(1); }
+
+  /// @brief Override the default sized allocation of qubits
+  /// here to be a bit more efficient than the default implementation
+  void addQubitsToState(std::size_t qubitCount,
+                        const void *stateDataIn = nullptr) override {
+    return;
+  }
+
+  /// @brief Reset the qubit state.
+  void deallocateStateImpl() override { stimCircuit.clear(); }
+
+  /// @brief Apply the noise channel on \p qubits
+  void applyNoiseChannel(const std::string_view gateName,
+                         const std::vector<std::size_t> &qubits) override {
+    // Do nothing if no execution context
+    if (!executionContext)
+      return;
+
+    // Do nothing if no noise model
+    if (!executionContext->noiseModel)
+      return;
+
+    // Get the name as a string
+    std::string gName(gateName);
+
+    // Cast size_t to uint32_t
+    std::vector<std::uint32_t> stimTargets;
+    stimTargets.reserve(qubits.size());
+    for (auto q : qubits)
+      stimTargets.push_back(static_cast<std::uint32_t>(q));
+
+    // Get the Kraus channels specified for this gate and qubits
+    auto krausChannels =
+        executionContext->noiseModel->get_channels(gName, qubits);
+
+    // If none, do nothing
+    if (krausChannels.empty())
+      return;
+
+    // TODO
+    return;
+  }
+
+  void applyGate(const GateApplicationTask &task) override {
+    std::string gateName(task.operationName);
+    std::transform(gateName.begin(), gateName.end(), gateName.begin(),
+                   ::toupper);
+    std::vector<std::uint32_t> stimTargets;
+
+    // These CUDA-Q rotation gates have the same name as Stim "reset" gates.
+    // Stim is a Clifford simulator, so it doesn't actually support rotational
+    // gates. Throw exceptions if they are encountered here.
+    if (gateName == "RX" || gateName == "RY" || gateName == "RZ")
+      throw std::runtime_error(fmt::format(
+          "Gate not supported by simulator: {}", task.operationName));
```

Comment:
nit: should we add a TODO comment here about the possibility of handling `+/- pi/2` rotations for future references?

Reply:
Updated in 20343fd07